include/rmm/mr/device/per_device_resource.hpp

```diff
@@ -0,0 +1,165 @@
+ *
+ * Returns a pointer to the resource set for the specified device. The initial resource is a
+ * `cuda_memory_resource`.
+ *
+ * This function is thread-safe.
+ *
+ * @param id The id of the target device
+ * @return Pointer to the current resource for `id`
+ */
+inline device_memory_resource* get_per_device_resource(cuda_device_id id)
+{
+  std::lock_guard<std::mutex> lock{detail::map_lock()};
+  auto& map = detail::get_map();
+  // If a resource was never set for `id`, set to the initial resource
+  return (map.find(id.value()) == map.end()) ? (map[id.value()] = detail::initial_resource())
+                                             : map[id.value()];
+}
+
+/**
+ * @brief Update the resource for the specified device
+ *
+ * If `new_mr` is not `nullptr`, sets the resource pointer for the device specified by `id` to
+ * `new_mr`. Otherwise, resets `id`s resource to the initial `cuda_memory_resource`.
+ *
+ * The object pointed to by `new_mr` must outlive the last use of the resource, else behavior is
+ * undefined. It is the caller's responsibility to maintain the lifetime of the resource object.
+ *
+ * This function is thread-safe.
+ *
+ * @param new_mr If not `nullptr`, pointer to new `device_memory_resource` to use as new resource
+ * for `id`
+ * @return Pointer to the previous the resource for `id`
+ */
+inline device_memory_resource* set_per_device_resource(cuda_device_id id,
+                                                       device_memory_resource* new_mr)
+{
+  std::lock_guard<std::mutex> lock{detail::map_lock()};
+  auto& map          = detail::get_map();
+  auto const old_itr = map.find(id.value());
+  // If a resource didn't previously exist for `id`, return pointer to initial_resource
+  auto old_mr     = (old_itr == map.end()) ? detail::initial_resource() : old_itr->second;
+  map[id.value()] = (new_mr == nullptr) ? detail::initial_resource() : new_mr;
+  return old_mr;
+}
+
+/**
+ * @brief Get the resource for the current device
+ *
+ * Returns a pointer to the resource set for the current device. The initial resource is a
+ * `cuda_memory_resource`.
+ *
+ * The "current device" is the device returned by `cudaGetDevice`.
+ *
+ * This function is thread-safe
+ *
+ * @return Pointer to the resource for the current device
+ */
+inline device_memory_resource* get_current_device_resource()
+{
+  return get_per_device_resource(detail::current_device());
+}
+
+/**
+ * @brief Update the resource for the current device
+ *
+ * If `new_mr` is not `nullptr`, sets the resource pointer for the current device to
+ * `new_mr`. Otherwise, resets the resource to the initial `cuda_memory_resource`.
+
+ * The "current device" is the device returned by `cudaGetDevice`.
+ *
+ * The object pointed to by `new_mr` must outlive the last use of the resource, else behavior is
+ * undefined. It is the caller's responsibility to maintain the lifetime of the resource object.
+ *
+ * This function is thread-safe.
```

Comment:
Is this really thread-safe? If another thread called `cudaSetDevice()` after this thread called `detail::current_device()` but before calling `set_per_device_resource` with the result, then you would end up setting the resource for the wrong device.

So it may be thread-safe wrt other threads calling this same function, but not wrt other threads setting the current CUDA device.

Reply:
`cudaSetDevice()` is specific to the calling thread, so one thread calling `cudaSetDevice()` does not impact another thread.

That said, I can and should be more specific about what other functions this is safe to be called concurrently with.

Reply:
Updated to say:
```
 * This function is thread-safe with respect to concurrent calls to `set_per_device_resource`,
 * `get_per_device_resource`, `get_current_device_resource`, and `set_current_device_resource`.
 * Concurrent calls to any of these functions will result in a valid state, but the order of
 * execution is undefined
```

Reply:
Ah, never noticed that about cudaSetDevice. I thought it was a context-level thing rather than thread-level.